config/assert_test.go

```diff
@@ -0,0 +1,209 @@
+		want       config.Assert
+		wantErr    error
+	}
+
+	expr := func(s string) hhcl.Expression {
+		return test.NewExpr(t, s)
+	}
+
+	tcases := []testcase{
+		{
+			name: "using literals",
+			assert: hcl.AssertConfig{
+				Assertion: expr(`"a" == "terramate"`),
+				Message:   expr(`"something"`),
+			},
+			want: config.Assert{
+				Assertion: false,
+				Message:   "something",
+			},
+		},
+		{
+			name: "accessing namespace values",
+			namespaces: namespaces{
+				"ns": nsvalues{
+					"a": "terramate",
+				},
+				"ns2": nsvalues{
+					"msg": "message",
+				},
+			},
+			assert: hcl.AssertConfig{
+				Assertion: expr(`ns.a == "terramate"`),
+				Message:   expr(`ns2.msg`),
+			},
+			want: config.Assert{
+				Assertion: true,
+				Message:   "message",
+			},
+		},
+		{
+			name: "warning defined",
+			namespaces: namespaces{
+				"warning": nsvalues{
+					"val": true,
+				},
+			},
+			assert: hcl.AssertConfig{
+				Assertion: expr(`true`),
+				Message:   expr(`"msg"`),
+				Warning:   expr("warning.val"),
+			},
+			want: config.Assert{
+				Assertion: true,
+				Message:   "msg",
+				Warning:   true,
+			},
+		},
+		{
+			name: "assertion undefined fails",
+			assert: hcl.AssertConfig{
+				Message: expr(`"something"`),
+			},
+			wantErr: errors.E(config.ErrSchema),
+		},
+		{
+			name: "message undefined fails",
+			assert: hcl.AssertConfig{
+				Assertion: expr(`true`),
+			},
+			wantErr: errors.E(config.ErrSchema),
+		},
+		{
+			name: "assertion is not boolean fails",
+			assert: hcl.AssertConfig{
+				Assertion: expr(`[]`),
+				Message:   expr(`"something"`),
+			},
+			wantErr: errors.E(config.ErrSchema),
+		},
+		{
+			name: "assertion eval fails",
+			assert: hcl.AssertConfig{
+				Assertion: expr(`unknown.access`),
+				Message:   expr(`"something"`),
+			},
+			wantErr: errors.E(eval.ErrEval),
+		},
+		{
+			name: "message is not string fails",
+			assert: hcl.AssertConfig{
+				Assertion: expr(`true`),
+				Message:   expr(`false`),
+			},
+			wantErr: errors.E(config.ErrSchema),
+		},
+		{
+			name: "message eval fails",
+			assert: hcl.AssertConfig{
+				Assertion: expr(`true`),
+				Message:   expr(`access.unknown`),
+			},
+			wantErr: errors.E(eval.ErrEval),
+		},
+		{
+			name: "warning is not boolean fails",
+			assert: hcl.AssertConfig{
+				Assertion: expr(`true`),
+				Message:   expr(`"msg"`),
+				Warning:   expr("[]"),
+			},
+			wantErr: errors.E(config.ErrSchema),
+		},
+		{
+			name: "warning eval fails",
+			assert: hcl.AssertConfig{
+				Assertion: expr(`true`),
+				Message:   expr(`"msg"`),
+				Warning:   expr("access.warning"),
+			},
+			wantErr: errors.E(eval.ErrEval),
+		},
+		{
+			name: "multiple errors",
+			assert: hcl.AssertConfig{
+				Assertion: expr(`unknown.val`),
+				Message:   expr(`false`),
+				Warning:   expr("access.warning"),
+			},
+			wantErr: errors.L(
+				errors.E(eval.ErrEval),
+				errors.E(config.ErrSchema),
+				errors.E(eval.ErrEval),
+			),
+		},
```

Comment:
```suggestion
		},
		{
			name: "using funcalls",
			assert: hcl.AssertConfig{
				Assertion: expr(`"A" == tm_upper("a")`),
				Message:   expr(`tm_upper("func")`),
			},
			want: config.Assert{
				Assertion: true,
				Message:   "FUNC",
			},
		},
```
I think a funcall test is missing.

Reply:
Noice, thanks :100: